calico/openstack/t_etcd.py

```diff
@@ -34,16 +49,283 @@ def __init__(self, driver, logger):
+                          json.dumps(self.profile_tags(profile_id)))
+
+    def profile_rules(self, profile_id):
+        inbound = []
+        outbound = []
+        for sgid in self.profile_tags(profile_id):
+            for rule in self.sgs[sgid]['security_group_rules']:
+                LOG.info("Neutron rule  %s : %s", profile_id, rule)
+
+                ethertype = rule['ethertype']
+                etcd_rule = {}
+
+                # Map the protocol field from Neutron to etcd format.
+                if rule['protocol'] is None:
+                    pass
+                elif rule['protocol'] == 'icmp':
+                    etcd_rule['protocol'] = {'IPv4': 'icmp',
+                                             'IPv6': 'icmpv6'}[ethertype]
+                    etcd_rule['icmp_type'] = rule['port_range_min']
+                else:
+                    etcd_rule['protocol'] = rule['protocol']
+
+                # OpenStack (sometimes) represents 'any IP address' by setting
+                # both 'remote_group_id' and 'remote_ip_prefix' to None.  We
+                # translate that to an explicit 0.0.0.0/0 (for IPv4) or ::/0
+                # (for IPv6).
+                net = rule['remote_ip_prefix']
+                if not (net or rule['remote_group_id']):
+                    net = {'IPv4': '0.0.0.0/0',
+                           'IPv6': '::/0'}[ethertype]
+
+                # src/dst_ports is a list in which each entry can be a single
+                # number, or a string describing a port range.
+                if rule['port_range_min'] == -1:
+                    port_spec = ['1:65535']
+                elif rule['port_range_min'] == rule['port_range_max']:
+                    if rule['port_range_min'] is not None:
+                        port_spec = [rule['port_range_min']]
+                    else:
+                        port_spec = None
+                else:
+                    port_spec = ['%s:%s' % (rule['port_range_min'],
+                                            rule['port_range_max'])]
+
+                # Put it all together and add to either the inbound or the
+                # outbound list.
+                if rule['direction'] == 'ingress':
+                    etcd_rule['src_tag'] = rule['remote_group_id']
+                    etcd_rule['src_net'] = net
+                    etcd_rule['src_ports'] = port_spec
+                    inbound.append(etcd_rule)
+                    LOG.info("=> Inbound Calico rule %s" % etcd_rule)
+                else:
+                    etcd_rule['dst_tag'] = rule['remote_group_id']
+                    etcd_rule['dst_net'] = net
+                    etcd_rule['dst_ports'] = port_spec
+                    outbound.append(etcd_rule)
+                    LOG.info("=> Outbound Calico rule %s" % etcd_rule)
+
+        return {'inbound_rules': inbound, 'outbound_rules': outbound}
+
+    def profile_tags(self, profile_id):
+        return profile_id.split('_')
```

Comment:
Methods like this (that don't use `self` anywhere in their body should probably be moved to being functions instead.